### PR TITLE
[Feature] Add HRSC Dataset

### PR DESCRIPTION
**Pull request by liuyanyi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Add HRSC dataset.
Part of https://github.com/open-mmlab/mmrotate/pull/54

## Modification
Implemented HRSCDataset based on XMLDataset in mmdet and HRSCDataset in OBBDetection.

Update performance:
| model | backbone | angle | mAP |
| ---| --- | --- | --- |
| rotated retinanet | r-50 | le90 | 72.60 |
| s2anet | r50 | le135 | 89.75 |

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zytx121** [comment]
Nice work!
Maybe we should add docs in `mmrotate/tools/data` to tell beginners how to prepare HRSD data sets.

**Comment by liuyanyi** [comment]
> @liuyanyi Nice work! Maybe we should add docs in `mmrotate/tools/data` to tell beginners how to prepare HRSD data sets.

I add doc for hrsc, but i can't find the original download link for dataset, i can only find dataset in [baidu ai studio](https://aistudio.baidu.com/aistudio/datasetdetail/31232).

File: configs/_base_/datasets/hrsc.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        type=dataset_type,
+        classwise=False,
+        ann_file=data_root + 'ImageSets/trainval.txt',
+        ann_subdir=data_root + 'FullDataSet/Annotations/',
+        img_subdir=data_root + 'FullDataSet/AllImages/',
+        pipeline=test_pipeline),
+    test=dict(
+        type=dataset_type,
+        classwise=False,
+        ann_file=data_root + 'ImageSets/trainval.txt',
```
Comment (yangxue0827) [inline configs/_base_/datasets/hrsc.py:50]: The test path is more appropriate. The same goes for other configuration files.

**Comment by yangxue0827** [comment]
It would be better if it could provide the corresponding performance.

File: configs/_base_/datasets/hrsc.py (hunk 1 above)
Comment (liuyanyi) [reply configs/_base_/datasets/hrsc.py]: fixed

**Comment by liuyanyi** [comment]
> It would be better if it could provide the corresponding performance.

Sure, I'll test the performance and add more config for hrsc.

File: configs/rotated_retinanet/README.md [hunk 2]
```diff
@@ -21,6 +21,12 @@ DOTA1.0
...
 | ResNet50 (1024,1024,200) | 76.50 | le90 | 1x |   | 16.7 | MS+RR | 2 | [rotated_retinanet_obb_r50_fpn_1x_dota_ms_rr_le90](./rotated_retinanet_obb_r50_fpn_1x_dota_ms_rr_le90.py) |  [model](https://download.openmmlab.com/mmrotate/v0.1.0/rotated_retinanet/rotated_retinanet_obb_r50_fpn_1x_dota_ms_rr_le90/rotated_retinanet_obb_r50_fpn_1x_dota_ms_rr_le90-1da1ec9c.pth) &#124; [log](https://download.openmmlab.com/mmrotate/v0.1.0/rotated_retinanet/rotated_retinanet_obb_r50_fpn_1x_dota_ms_rr_le90/rotated_retinanet_obb_r50_fpn_1x_dota_ms_rr_le90_20220210_114843.log.json)
 
+HRSC
+
+|       Backbone        |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | Aug | Batch Size |                                            Configs                                            | Download |
+|:---------------------:|:-----:|:-----------:|:---------:|:---------:|:---------:|:---------:|:---------:|:---------------------------------------------------------------------------------------------:|:-------------:|
+|  ResNet50 (800,800)   | 72.60 | le90 | 1x |  |  | - | 2 | [rotated_retinanet_obb_r50_fpn_3x_hrsc_le90](./rotated_retinanet_obb_r50_fpn_3x_hrsc_le90.py) |
```
Comment (yangxue0827) [inline configs/rotated_retinanet/README.md:28]: lr schd: 1x -> 3x ?

Comment (liuyanyi) [reply configs/rotated_retinanet/README.md:28]: Yes, That's a typo, i'll fix that.

**Review by yangxue0827 — APPROVED** [review]
LGTM

File: mmrotate/datasets/hrsc.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                    gt_headers_ignore, dtype=np.float32)
+            else:
+                data_info['ann']['bboxes_ignore'] = np.zeros((0, 5),
+                                                             dtype=np.float32)
+                data_info['ann']['labels_ignore'] = np.array([],
+                                                             dtype=np.int64)
+                data_info['ann']['polygons_ignore'] = np.zeros(
+                    (0, 8), dtype=np.float32)
+                data_info['ann']['headers_ignore'] = np.zeros((0, 2),
+                                                              dtype=np.float32)
```
Comment (jbwang1997) [inline mmrotate/datasets/hrsc.py:184]: Seems no ignored objects are added.

File: mmrotate/datasets/hrsc.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        eval_results = OrderedDict()
+        iou_thrs = [iou_thr] if isinstance(iou_thr, float) else iou_thr
+        if metric == 'mAP':
+            assert isinstance(iou_thrs, list)
+            mean_aps = []
+            for iou_thr in iou_thrs:
+                print_log(f'\n{"-" * 15}iou_thr: {iou_thr}{"-" * 15}')
+                mean_ap, _ = eval_map(
+                    results,
+                    annotations,
```
Comment (jbwang1997) [inline mmrotate/datasets/hrsc.py:242]: The name of `eval_map` has been changed to `eval_rbbox_map`.

Comment (liuyanyi) [reply mmrotate/datasets/hrsc.py]: fix

File: mmrotate/datasets/hrsc.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+                            dtype=np.float32)
+                    else:
+                        bbox = bbox[0, :-1]
+                    head = np.array([
+                        int(obj.find('header_x').text),
+                        int(obj.find('header_y').text)
+                    ],
+                                    dtype=np.int64)
+                except:  # noqa: E722
+                    continue
```
[the remark this answers is not in the document]
Comment (liuyanyi) [reply mmrotate/datasets/hrsc.py]: it used for debug, already removed

File: mmrotate/datasets/hrsc.py (hunk 3 above)
Comment (liuyanyi) [reply mmrotate/datasets/hrsc.py:180]: i just copied that from dota.py, all the ignored field is empty. Should I remove the if branch?

Comment (jbwang1997) [reply mmrotate/datasets/hrsc.py:178]: Keep this branch is okay.****

**Review by jbwang1997 — APPROVED** [review]
LGTM

File: configs/rotated_retinanet/README.md (hunk 2 above)
Comment (yangxue0827) [reply configs/rotated_retinanet/README.md]: Hello, I have sent an invitation to maintain mmrotate to your 163 mailbox (obtained from github). If you are interested, you can contact me through wechat in the email. @liuyanyi

**Review by zytx121 — APPROVED** [review]
LGTM

**Pull request merged**